[all] Use NVI design pattern for drawVisual

damienmarchal opened the pull request:
Actually what we called "delegate" pattern (the one we used to think about to fix the "call-super" anti-pattern) is often refereed on intenret as Non Virtual Interface (a pattern very similar to the Template Method Pattern). 


Here is a attempt to see it in use on VisualModel::drawVisual; 



______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

damienmarchal committed:
Implement the drawVisual method using the NVI pattern to factorize access control.

damienmarchal committed:
Propagate the change.

damienmarchal committed:
FIXUP

damienmarchal committed:
[Sofa.Core] Refactor VisualVisitor::processVisualModel to factorize code;

Signed-off-by: Damien Marchal <damien.marchal@univ-lille1.fr>

damienmarchal committed:
FIXUP

damienmarchal committed:
Adds a transitional "virtual" keyword on drawVisual to not break third party plugin.

alxbilger commented on Sofa/framework/Core/src/sofa/core/visual/VisualModel.cpp:
```diff
@@ -21,10 +21,34 @@
+    if( d_componentState.getValue() == sofa::core::objectmodel::ComponentState::Invalid )
+        return;
+
+    // don't draw if this component is specifically configure to be disabled
```
```suggestion
    // don't draw if this component is specifically configured to be disabled
```

alxbilger commented on Sofa/framework/Simulation/Core/src/sofa/simulation/VisualVisitor.cpp:
```diff
@@ -1,24 +1,24 @@
-*                                                                             *
-* Contact information: contact@sofa-framework.org                             *
-******************************************************************************/
+ *                 SOFA, Simulation Open-Framework Architecture                *
```
extra spaces

alxbilger commented on Sofa/framework/Simulation/Core/src/sofa/simulation/VisualVisitor.cpp:
```diff
@@ -96,47 +108,51 @@ void VisualDrawVisitor::bwdVisualModel(simulation::Node* /*node*/,core::visual::
+        return;
 
-        vm->drawVisual(vparams);
+    // don't draw if this component is specifically configure to be disabled
```
```suggestion
    // don't draw if this component is specifically configured to be disabled
```

damienmarchal committed:
Update Sofa/framework/Core/src/sofa/core/visual/VisualModel.cpp

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

damienmarchal committed:
Update Sofa/framework/Simulation/Core/src/sofa/simulation/VisualVisitor.cpp

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

damienmarchal commented:
@alxbilger  thank for the review.

damienmarchal committed:
FIX alignement

damienmarchal committed:
Merge remote-tracking branch 'cristal/pr-use-pattern-nvi-for-drawvisual' into pr-use-pattern-nvi-for-drawvisual

damienmarchal commented:
Do everyone agrees on having the conditions I moved up into the NVI VisualModel::drawVisual()  and with the one I added into the VisualVisitor ? 

Eg: 
```cpp
void VisualModel::drawVisual(const VisualParams* vparams)
{
    // don't draw if specified not to do so in the user interface
    if (!vparams->displayFlags().getShowVisualModels())
        return;

    // don't draw if the component is not in valid state (we can change the condition so Loading/Unknow states are drawn differently) 
    if( d_componentState.getValue() == sofa::core::objectmodel::ComponentState::Invalid )
        return;

    // don't draw if this component is specifically configured to be disabled (It is questionnable to have a data for that). 
    // having a d_enable on Base would be better. 
    if (!d_draw.getValue())
        return;

    // RAI to be sure the states changed during the rendering don't leaks 
    const auto stateLifeCycle = vparams->drawTool()->makeStateLifeCycle();

    doDrawVisual(vparams);
}
```

alxbilger commented on Sofa/framework/Core/src/sofa/core/visual/VisualModel.h:
```diff
@@ -48,9 +48,25 @@ class SOFA_CORE_API VisualModel : public virtual objectmodel::BaseObject
+    /**
+     *  \brief Display the VisualModel object.
+     *
+     *  TODO(dmarchal, 2023-06-09): Deprecate VI et use NVI design pattern: In one year, remove the virtual keyword so that everyone
```
```suggestion
     *  TODO(dmarchal, 2023-06-09): Deprecate VI and use NVI design pattern: In one year, remove the virtual keyword so that everyone
```

damienmarchal committed:
Update Sofa/framework/Core/src/sofa/core/visual/VisualModel.h

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

alxbilger commented on applications/plugins/SofaSphFluid/src/SofaSphFluid/OglFluidModel.inl:
```diff
@@ -584,12 +584,8 @@ void OglFluidModel<DataTypes>::drawSprites(const core::visual::VisualParams* vpa
 {
-    if (!vparams->displayFlags().getShowVisualModels()) return;
-
-    vparams->drawTool()->saveLastState();
```
You forgot to remove `vparams->drawTool()->restoreLastState();` at the end of the function

alxbilger commented on Sofa/GL/Component/Rendering2D/src/sofa/gl/component/rendering2d/OglLabel.h:
```diff
@@ -49,12 +49,11 @@ class SOFA_GL_COMPONENT_RENDERING2D_API OglLabel : public core::visual::VisualMo
     Data<sofa::type::RGBAColor> d_color; ///< The color of the text to display. (default='gray')
     Data<bool>                   d_selectContrastingColor ; ///< Overide the color value but one that contrast with the background color
     Data<unsigned int>           d_updateLabelEveryNbSteps; ///< Update the display of the label every nb of time steps
-    Data<bool>                   d_visible; ///< Is label displayed
```
DeprecatedAndRemoved ?

damienmarchal replied:
If we move forward for https://github.com/sofa-framework/sofa/pull/3934 I will use it :)

damienmarchal replied:
I disagree @hugtalbot as it shouldn't be removed in this PR.

alxbilger replied:
I am confused. What do you plan to do with this Data? Currently, you have removed the class member and added an alias to `d_draw`. But, according to https://www.sofa-framework.org/community/doc/contributing-to-sofa/lifecycle-mechanisms/, when deleting a Data, its type must be `DeprecatedAndRemoved` to highlight the fact that a variable with this name existed, but has been removed.

damienmarchal replied:
I plan to follow the lifecycle, so before deleting something, first deprecate it.

damienmarchal replied:
I plan to follow the lifecycle, so before deleting something, first deprecate it.

damienmarchal committed:
FIXUP

alxbilger commented on Sofa/framework/Core/src/sofa/core/visual/VisualModel.cpp:
```diff
@@ -21,10 +21,34 @@
+
+    const auto stateLifeCycle = vparams->drawTool()->makeStateLifeCycle();
+
+    doDrawVisual(vparams);
```
```suggestion
    if (vparams->displayFlags().getShowWireFrame())
    {
        vparams->drawTool()->setPolygonMode(0, true);
    }

    doDrawVisual(vparams);
```
What do you think about enforcing wireframe rendering at the VisualModel level?

damienmarchal replied:
It is a good idea that naturally fit into the generalized deployement of NVI. Do you want me to add it in this PR ?

damienmarchal commented on Sofa/framework/Core/src/sofa/core/visual/VisualModel.h:
```diff
@@ -48,9 +48,25 @@ class SOFA_CORE_API VisualModel : public virtual objectmodel::BaseObject
     SOFA_ABSTRACT_CLASS(VisualModel, objectmodel::BaseObject);
     SOFA_BASE_CAST_IMPLEMENTATION(VisualModel)
+
+    Data<bool> d_draw; ///< Display the visual model or not
```
I'm not sure.

VisualFlag is just a filter on what type of component is displayed (all VisualModel) and how (eg: wireframe)... precise control of which object are rendered has nothing to do with that. 
To me the introduced d_draw, is the same as is the "drawObject" data we have in MechanicalObject (which actually is very very useful and nothing related with the visual flags);

That's say I consider that for the specific case of visual model "d_draw" may be redundant with a d_enable<bool> implemented in BaseObject that activate/deactive a component (which in the specific case of a visual model should disable its rendering).  But there is no such "d_enable" data in BaseObject.  I'm open to rename d_draw to d_enable if we want to prepare compatibility for such "enable" boolean to activate/deactive component.

damienmarchal committed:
Add proper deprecation of "visible" flag in OglLabel

damienmarchal committed:
FIXUP

damienmarchal committed:
Implement top level control of drawWireFrame

alxbilger commented:
[ci-build][with-all-tests]

alxbilger commented on Sofa/framework/Core/src/sofa/core/visual/VisualModel.cpp:
```diff
@@ -21,10 +21,39 @@
+
+    // don't draw if the component is not in valid state
+    if( d_componentState.getValue() == sofa::core::objectmodel::ComponentState::Invalid )
+        return;
```
For that, all component must set it to valid at some point, otherwise it is `UNDEFINED`

damienmarchal replied:
As pointed by @alxbilger, using isComponentStatevalid() is not the same. The condition is really != Invalid to take into account the classes inheriting from VisualModel but are not setting their component state;

damienmarchal commented:
CI passed...

damienmarchal committed:
[all] Rename d_draw by d_enable VisualModel to make it consistent with other components

alxbilger reviewed (changes requested):
Now that you have renamed `d_draw` into `d_enable`, you need to deprecate `d_draw` in the derived classes instead of removing them.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-use-pattern-nvi-for-drawvisual

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-use-pattern-nvi-for-drawvisual

damienmarchal committed:
Use the newly introduced RemovedData to manage the use of "enable" instea of "draw" and "visible"

damienmarchal committed:
Test for d_enable before d_componentState == Valid to avoid updating the component if disable.

damienmarchal committed:
Fix OglLable_test to remove the "visible" attribute;

alxbilger commented on Sofa/framework/Config/src/sofa/config.h.in:
```diff
@@ -244,6 +244,11 @@ class DeprecatedAndRemoved {};
+    SOFA_ATTRIBUTE_DEPRECATED( \
+        "v23.06 (PR#3931)", "v23.12", \
+        "Attribute 'visible' is deprecated, use 'draw' to control rendering.")
+
```
```suggestion
```

I think this is no longer used

damienmarchal commented on Sofa/framework/Core/src/sofa/core/visual/VisualModel.cpp:
```diff
@@ -21,10 +21,39 @@
+        return;
+
+    // don't draw if the component is not in valid state
+    if( d_componentState.getValue() == sofa::core::objectmodel::ComponentState::Invalid )
+        return;
```
So ? It looks like you are making a suggestion of change but saying it will not work the same.

damienmarchal committed:
Update Sofa/framework/Config/src/sofa/config.h.in

Co-authored-by: Alex Bilger <alxbilger@users.noreply.github.com>

alxbilger commented:
[ci-build][with-all-tests]
